java-checks/src/main/java/org/sonar/java/checks/PreparedStatementLoopInvariantCheck.java

```diff
@@ -0,0 +1,210 @@
+    });
+  }
+
+  private void reportIfLoopInvariant(Set<String> declaredOrAssignedLocals, Candidate candidate) {
+    if (isLoopInvariant(declaredOrAssignedLocals, candidate)) {
+      var secondaryLocation = new JavaFileScannerContext.Location(
+        "Enclosing loop",
+        Objects.requireNonNull(candidate.enclosingLoop)
+      );
+      reportIssue(
+        candidate.invocation,
+        "Move this loop-invariant setter invocation out of this loop.",
+        List.of(secondaryLocation),
+        null
+      );
+    }
+  }
+
+  private static boolean isLoopInvariant(Set<String> declaredOrAssignedLocals, Candidate candidate) {
+    return (candidate.identifierArguments.stream().noneMatch(declaredOrAssignedLocals::contains));
+  }
+
+  private static Candidate getCandidate(MethodInvocationTree invocation) {
+    var identifierArguments = new ArrayList<String>();
+    for (var arg : invocation.arguments()) {
+      if (arg.is(Tree.Kind.IDENTIFIER)) {
+        identifierArguments.add(((IdentifierTree) arg).name());
+      } else {
+        var argValue = ExpressionUtils.resolveAsConstant(arg);
+        if (argValue == null) return null;
+      }
+    }
+
+    var loop = findEnclosingLoop(invocation);
+    if (loop == null) {
+      return null;
+    }
+    return new Candidate(invocation, identifierArguments, loop);
+  }
+
+  private static StatementTree findEnclosingLoop(Tree tree) {
+    while (!Objects.requireNonNull(tree).is(Tree.Kind.METHOD)) {
+      if (isLoop(tree)) {
+        return (StatementTree) tree;
+      }
+      tree = tree.parent();
+    }
+    return null;
+  }
+
+  private static boolean isLoop(Tree tree) {
+    switch (tree.kind()) {
+      case FOR_STATEMENT:
+      case FOR_EACH_STATEMENT:
+      case WHILE_STATEMENT:
+      case DO_STATEMENT:
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  private static class Candidate {
+    public final MethodInvocationTree invocation;
+    public final List<String> identifierArguments;
+    public final StatementTree enclosingLoop;
+
+    private Candidate(MethodInvocationTree invocation, List<String> argumentVariables, StatementTree enclosingLoop) {
+      this.invocation = invocation;
+      this.identifierArguments = argumentVariables;
+      this.enclosingLoop = enclosingLoop;
+    }
+  }
+
+  private static class MethodInvocationCollector extends BaseTreeVisitor {
+
+    public final List<MethodInvocationTree> invocations = new ArrayList<>();
+    private final MethodMatchers matchers;
+
+    public MethodInvocationCollector(MethodMatchers matchers) {
+      this.matchers = matchers;
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      if (matchers.matches(tree)) {
+        invocations.add(tree);
+      }
+      super.visitMethodInvocation(tree);
+    }
+  }
+
+  private static class DeclaredOrAssignedLocalsCollector extends BaseTreeVisitor {
```

Comment:
Yes, that's a good idea, I think this could indeed be useful for similar situations. I think we should put it into a separate PR maybe? (we could put this into ExpressionUtils or somewhere else and need additional tests for it, so I see it as a separate task)